### ISSUE TITLE
Weighted directed label propagation gets stuck (or is very slow)

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2561]
This was found by the fuzzer. Label propagation running on directed weighted graphs can time out even on some small graphs. It gets stuck in some part of the code that is not even interruptible.

https://oss-fuzz.com/testcase-detail/5755444401340416
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=67840

We need to distill this down to a manageable size testcase.

**Comment by szhorvat** [comment]
@vtraag Would you be able to look at this if I produce a small testcase that is in a convenient format?

**Comment by szhorvat** [comment]
I'm guessing that the labels must be running around in circles indefinitely. Something about the graph structure and choice of weights makes it so that the probability of reaching a stable state is very low, or perhaps it is plainly impossible to have a stable state.

@vtraag Do you know if directed label propagation has been studied or whether this is an ad-hoc extension in igraph? Are there any results on whether the propagation procedure will stabilize in the directed case? I fear that fixing this may not be feasible ...  Perhaps the best approach is to make the function interruptible and document it as a potential issue? I'll leave it to you to investigate and make a decision since you've recently worked with label propagation.

**Comment by vtraag** [comment]
I'll look at it later. Perhaps @lovre has some useful input?

**Comment by ntamas** [comment]
> I'm guessing that the labels must be running around in circles indefinitely.

As far as I know this is a possibility when the labels are updated synchronously, but an easy way out is to update the labels asynchronously and to use a different, randomized update order in every cycle. But it seems like it doesn't work in this case; this is a cycle I'm getting:

```
3 7 3 3 7 3 7 7 3 10
3 7 3 3 7 10 7 7 3 10
3 7 3 3 7 10 7 7 10 10
10 10 3 3 3 3 7 7 10 10
3 10 3 3 3 3 7 7 3 10
```

**Comment by ntamas** [comment]
There are two concerns with the current label propagation implementation:

1. It is not interruptible at all -- I think we should simply place an `IGRAPH_ALLOW_INTERRUPTION()` call in every 1000 iteration or so.
2. The implementation seems to call RNG_BEGIN() and RNG_END() for every iteration; I think these should be moved outside the main loop so they get called only once.

**Comment by szhorvat** [comment]
>     2. The implementation seems to call RNG_BEGIN() and RNG_END() for every iteration; I think these should be moved outside the main loop so they get called only once.

The problem is that we also call `vector_shuffle()` within the main loop, which already includes an `RNG_BEGIN/END` pair. Such pairs shouldn't be nested, as they reduce randomness by resetting the random state unnecessarily.

Ideally, we could get rid of all the RNG_BEGIN/END for release 1.0, and move them to the R interface.

**Comment by ntamas** [comment]
> The problem is that we also call vector_shuffle() within the main loop, which already includes an RNG_BEGIN/END pair. 

aaah. This is something that should be fixed in 1.0 (if we decide not to get rid of these calls -- it might prove to be more tricky than expected). IMHO if we decide to keep these, then only top-level igraph functions that are not expected to be called by other top-level igraph functions should ever call `RNG_BEGIN()` and `RNG_END()`, so `igraph_vector_shuffle()` should not ever call it.

**Comment by vtraag** [comment]
Perhaps best to wait to merge my fast label propagation PR before diving in further?

**Comment by szhorvat** [comment]
I added interruption support, can you please resolve conflicts and merge #2451, @vtraag, if you think it's ready?

**Comment by szhorvat** [comment]
Now that #2451 is merged, this is a good time to come back to this issue. Do you have any ideas here @vtraag ?

**Comment by szhorvat** [comment]
@vtraag, any suggestions on how to address this?

**Comment by vtraag** [comment]
No particular direct suggestion. You did confirm that this particular problem still exists in the most recent version?

**Comment by szhorvat** [comment]
> You did confirm that this particular problem still exists in the most recent version?

Yes. See https://github.com/igraph/igraph/issues/2561#issuecomment-2038169153
